pkg/auth/authenticator/password/ldappassword/util/identityfactory.go

```diff
@@ -0,0 +1,70 @@
+package ldaputil
+
+import (
+	"fmt"
+
+	"github.com/go-ldap/ldap"
+
+	"k8s.io/kubernetes/pkg/util"
+
+	authapi "github.com/openshift/origin/pkg/auth/api"
+)
+
+// LDAPUserIdentityFactory creates Identites for LDAP user entries.
+type LDAPUserIdentityFactory interface {
+	IdentityFor(user *ldap.Entry) (identity authapi.UserIdentityInfo, err error)
+}
+
+// LDAPUserToIdentityMapping holds LDAP attributes that map to OpenShift Identity fields in oder to
+// create a deterministic mapping from LDAP user entries to OpenShift Identities
+type LDAPUserToIdentityMapping struct {
+	// EmailAttribute is the optional list of LDAP attributes to use for the email address of the user
+	// identity. The first attribute with a non-empty value is used.
+	EmailAttribute []string
+	// NameAttribute is the optional list of LDAP attributes to use for the display name of the user
+	// identity. The first attribute with a non-empty value is used.
+	NameAttribute []string
+	// PreferredUsernameAttribute is the optional list of LDAP attributes to use for the preferred
+	// username of the user identity. The first attribute with a non-empty value is used. If not
+	// specified, the id determined by AttributeID is used as the preferred login.
+	PreferredUsernameAttribute []string
+	// IDAttribute is the required list of LDAP attributes to use for the id address of the user identity.
+	// The first attribute with a non-empty value is used. If no attributes have values, login fails.
+	IDAttribute []string
+}
+
+func (m *LDAPUserToIdentityMapping) AllAttributes() util.StringSet {
+	attrs := util.NewStringSet(m.EmailAttribute...)
+	attrs.Insert(m.NameAttribute...)
+	attrs.Insert(m.PreferredUsernameAttribute...)
+	attrs.Insert(m.IDAttribute...)
+	return attrs
+}
+
+// DefaultLDAPUserIdentityFactory creates Identities for LDAP user entries using an LDAPUserToIdentityMapping
+type DefaultLDAPUserIdentityFactory struct {
+	ProviderName string
+	Mapping      LDAPUserToIdentityMapping
+}
+
+func (f *DefaultLDAPUserIdentityFactory) IdentityFor(user *ldap.Entry) (identity authapi.UserIdentityInfo, err error) {
+	uid := GetAttributeValue(user, f.Mapping.IDAttribute)
+	if uid == "" {
+		err = fmt.Errorf("Could not retrieve a non-empty value from %v attributes for dn=%q", f.Mapping.IDAttribute, user.DN)
+		return
+	}
+	id := authapi.NewDefaultUserIdentityInfo(f.ProviderName, uid)
+
+	// Add optional extra attributes if present
+	for k, attrs := range map[string][]string{
```

Comment:
I wasn't a fan the first time I review this and I'm still not a fan.  Can you turn thes into`.Email(*ldap.Entry) string` sort of methods the `LDAPUserToIdentityMapping` object.